### PR TITLE
core/rawdb: initial fix for fatal gap in chain error

### PR DESCRIPTION
**Pull request by meowsbits**
Rel https://github.com/etclabscore/core-geth/issues/8

Signed-off-by: meows <b5c6@protonmail.com>

### CODE REVIEW

**Comment by meowsbits** [comment]
Output currently:

```
> rm -rf /tmp/gethdd/geth/chaindata && cp -a /tmp/gethdd/geth/chaindata.bak /tmp/gethdd/geth/chaindata && make geth && ./build/bin/geth --datadir /tmp/get
hdd --mordor --verbosity 4                                                                                                                         
env GO111MODULE=on go run build/ci.go install ./cmd/geth                                                                                           
>>> /home/ia/go1.13.5.linux-amd64/go/bin/go install -ldflags -X main.gitCommit=c19f3406ee50e2c0128bd5e69fd70dcf4ebded30 -X main.gitDate=20200303 -v ./cmd/
geth                                                                                                                                      
github.com/ethereum/go-ethereum/core/rawdb                                                                                                                
github.com/ethereum/go-ethereum/core                                              
github.com/ethereum/go-ethereum/eth/filters                                                                                                               
github.com/ethereum/go-ethereum/light                                                                                                                     
github.com/ethereum/go-ethereum/eth/downloader                                      
github.com/ethereum/go-ethereum/internal/ethapi                                                           
github.com/ethereum/go-ethereum/graphql                                                                                                                   
github.com/ethereum/go-ethereum/node                                                                                                                     
github.com/ethereum/go-ethereum/accounts/abi/bind/backends                                                                                                
github.com/ethereum/go-ethereum/eth                                               
github.com/ethereum/go-ethereum/les                                                                   
github.com/ethereum/go-ethereum/cmd/utils                                                             
github.com/ethereum/go-ethereum/cmd/geth                                                                      
Done building.                                                                                                                 
Run "./build/bin/geth" to launch geth.                                                                                          
INFO [03-03|11:13:02.728] Bumping default cache on mainnet         provided=1024 updated=4096                                   
DEBUG[03-03|11:13:02.728] Sanitizing Go's GC trigger               percent=25                                                                     
INFO [03-03|11:13:02.730] Maximum peer count                       ETH=50 LES=0 total=50                                                          
INFO [03-03|11:13:02.730] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"                    
DEBUG[03-03|11:13:02.730] FS scan times                            list=20.996µs set=405ns diff=521ns                                                     INFO [03-03|11:13:02.732] Starting peer-to-peer node               instance=CoreGeth/v1.11.1-unstable-c19f3406-20200303/linux-amd64/go1.13.5      
INFO [03-03|11:13:02.732] Allocated trie memory caches             clean=1024.00MiB dirty=1024.00MiB                                                      INFO [03-03|11:13:02.732] Allocated cache and file handles         database=/tmp/gethdd/geth/chaindata cache=2.00GiB handles=524288
DEBUG[03-03|11:13:02.764] Chain freezer table opened               database=/tmp/gethdd/geth/chaindata/ancient table=bodies items=501958 size=2.75MiB     DEBUG[03-03|11:13:02.773] Chain freezer table opened               database=/tmp/gethdd/geth/chaindata/ancient table=receipts items=501958 size=1.48MiB
DEBUG[03-03|11:13:02.783] Chain freezer table opened               database=/tmp/gethdd/geth/chaindata/ancient table=diffs    items=501958 size=2.86MiB   DEBUG[03-03|11:13:02.874] Chain freezer table opened               database=/tmp/gethdd/geth/chaindata/ancient table=headers  items=501958 size=134.81MiB
DEBUG[03-03|11:13:02.892] Chain freezer table opened               database=/tmp/gethdd/geth/chaindata/ancient table=hashes   items=501958 size=15.32MiB
INFO [03-03|11:13:02.892] Opened ancient database                  database=/tmp/gethdd/geth/chaindata/ancient                                            WARN [03-03|11:13:02.898] Freezer/KV validation error, attempting freezer repair error="gap (chaindb=#839245 frozen=#501958) in the chain between ancients
 and leveldb"                                                                                                                                             WARN [03-03|11:13:02.898] Freezer repair OK                                        
WARN [03-03|11:13:02.898] Persistent Freezer/KV gap: Truncating KV database to freezer height ancients=501958 kv.head_header_number=839245 kv.head_header_hash=f2d626…e810c9                                                        
WARN [03-03|11:13:03.304] Removing KV block data                   n=830000 hash=0xa6c400d4ac4ef22addf56575cf79c2082577a00988e70c6e19c76b643baf0ba6       WARN [03-03|11:13:03.764] Removing KV block data                   n=820000 hash=0xdf812f64b9068c12fe1b83dd78e985433808e24b9ecab891166edf123ef8063d
WARN [03-03|11:13:04.198] Removing KV block data                   n=810000 hash=0xd925bb9d7ef68bc774c33026d3d033bd49038cd4ebc0eeacdfa278877e983856       WARN [03-03|11:13:04.649] Removing KV block data                   n=800000 hash=0xdd61074fdddd0ba9fa5c59ebf5e25ea807329a07587fda567ea5be8581b0a60a
WARN [03-03|11:13:05.109] Removing KV block data                   n=790000 hash=0x252ffed8c274a0980d49422b70b9fe7ec985ab12ef74212b09a20a5d2d89b10f
WARN [03-03|11:13:05.570] Removing KV block data                   n=780000 hash=0x7dbacff27c2e3b2dd305b3264e6e38c3ea4375025d852da99e0f9565aa367f82
WARN [03-03|11:13:06.035] Removing KV block data                   n=770000 hash=0xaba5c90a9c902967829ca05551d21a974c336740fe9d5191bb73f29e66018cee       WARN [03-03|11:13:06.496] Removing KV block data                   n=760000 hash=0x37125451ed0b0d2e06824ccfd52b04ed3b4b56a2403b1cc243dc180d5bf6e79a
WARN [03-03|11:13:06.951] Removing KV block data                   n=750000 hash=0x4b7a337d49ba8c3fa40fae4b1b51ffa101bc578615c0633b1e36c21c7e6fd41e       WARN [03-03|11:13:07.414] Removing KV block data                   n=740000 hash=0x0910ae3c6ea161148bb835bb4ce3fbea53f63759bced7d93b3cc984e8b9fd621
WARN [03-03|11:13:07.822] Removing KV block data                   n=730000 hash=0xe73117529fded7405786a22ef4bc7ee6359a2e54cd361353ce84f81ae0f3645f       WARN [03-03|11:13:08.182] Removing KV block data                   n=720000 hash=0x05c11768b81f00340cea324c7e7f662a48f8946a83c62ab24497e57212b6eb45
WARN [03-03|11:13:08.526] Removing KV block data                   n=710000 hash=0x5e2c9f44565bf028ebd9fde8346d1bfa30ec1dd2cf13313ccdbb768787519678       WARN [03-03|11:13:08.880] Removing KV block data                   n=700000 hash=0xcf5987b48846bf92802e35d9c435d915ceeeb5a68b7bab5b884881b05a0921dd
WARN [03-03|11:13:09.250] Removing KV block data                   n=690000 hash=0xb6f8baacee4f564029aff5638a09b54797f15a24b3b1540dd3c64440e4bf093b       WARN [03-03|11:13:09.605] Removing KV block data                   n=680000 hash=0x81e6f3be9ff819dbcdd80d7ac5fc73c7a7b81dbd5573f3d71699f74d52131523
WARN [03-03|11:13:09.950] Removing KV block data                   n=670000 hash=0xff3c2b100edc55ea00d1e29741772b9a36c244b09219742387715ec376f63924       WARN [03-03|11:13:10.295] Removing KV block data                   n=660000 hash=0x09bca32a68c46f0d3c90d83d648badbae5a04bb6931afd5fd89da0ccbc64043e
WARN [03-03|11:13:10.612] Removing KV block data                   n=650000 hash=0xc4cdd16ff8ee35697137a622966ef8365ea6d6232c4b3990afc93975f2236ef9
WARN [03-03|11:13:10.924] Removing KV block data                   n=640000 hash=0x391cd485c6430504698dc990e57bce9f4cd250bf1ba24ba21181ad919e319cf2
WARN [03-03|11:13:11.248] Removing KV block data                   n=630000 hash=0x2eed934410f220aaba8ed36b0c9d17ffed52ede3efd39f9a11a6f4e75dd45158
WARN [03-03|11:13:11.599] Removing KV block data                   n=620000 hash=0x8b3dfc97633c770d784874e8de5219c8a8e7c4232e1ace81f9f40e840326b5d9
WARN [03-03|11:13:11.925] Removing KV block data                   n=610000 hash=0xcbdd21b6e79dd4b4e844661aa0ffa6ea864aca03ac50e2c066d3c694419ef55d
WARN [03-03|11:13:12.261] Removing KV block data                   n=600000 hash=0x4c41542fb189ec3bf303a772f5e634073401069b9d1b21eaf73e495e029b5f9a
WARN [03-03|11:13:12.574] Removing KV block data                   n=590000 hash=0x774ce3ffc9ad6fb4ef8675bdf45ab20160c8ae1a1cf01659fc7443ee8805a56b
WARN [03-03|11:13:12.912] Removing KV block data                   n=580000 hash=0x393a8b1f394d09f06022053e912f5974986befd3ff4a8a2b368e144a0120bac0
WARN [03-03|11:13:13.247] Removing KV block data                   n=570000 hash=0xf3594daf24eb4c6da64b3e19a4a1e12a3e6327cde09ea8fe9583943e778e0d8a
WARN [03-03|11:13:13.554] Removing KV block data                   n=560000 hash=0x80e8a7e4895f577e7d0bbb7d329c831c215a23330edc3255237df6f2dab6e4a3
WARN [03-03|11:13:13.859] Removing KV block data                   n=550000 hash=0x42fdff6f256c1b412bb4d76e06569dd53d7eba18141f38169baf36b602932bb1
WARN [03-03|11:13:14.042] Finished KV truncation                                        
WARN [03-03|11:13:14.042] Writing KV head header                   hash=0xfc60ccf6933322f1bf8f81b570a3513ac21ed23e611e3388e7b0472d9f1aff3e           
INFO [03-03|11:13:14.043] Persisted trie from memory database      nodes=0 size=0.00B     time=2.149µs gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesi
ze=0.00B                                                                                                                                             
INFO [03-03|11:13:14.043] Found stored genesis block               config="NetworkID: 7, ChainID: 63 Engine: ethash EIP1014: 301243 EIP1052: 301243 EIP140: 0 EIP145: 301243 EIP150: 0 EIP155: 0 EIP160: 0 EIP161abc: 0 EIP161d: 0 EIP170: 0 EIP198: 0 EIP211: 0 EIP212: 0 EIP213: 0 EIP214: 0 EIP658: 0 EIP7: 0 Eth
ashECIP1017: 0 EthashECIP1041: 0 EthashEIP100B: 0 EthashEIP2: 0 EthashHomestead: 0 "                                                                   
DEBUG[03-03|11:13:14.043] Current full block not old enough        number=0 hash=a68ebd…59a3f1 delay=90000                  
INFO [03-03|11:13:14.043] Initialised chain configuration          config="NetworkID: 7, ChainID: 63 Engine: ethash EIP1014: 301243 EIP1052: 301243 EIP110
8: 999983 EIP1344: 999983 EIP140: 0 EIP145: 301243 EIP150: 0 EIP152: 999983 EIP155: 0 EIP160: 0 EIP161abc: 0 EIP161d: 0 EIP170: 0 EIP1884: 999983 EIP198:
0 EIP2028: 999983 EIP211: 0 EIP212: 0 EIP213: 0 EIP214: 0 EIP2200: 999983 EIP658: 0 EIP7: 0 EthashECIP1017: 0 EthashECIP1041: 0 EthashEIP100B: 0 EthashEIP
2: 0 EthashHomestead: 0 "                                                                                                                                 INFO [03-03|11:13:14.043] Disk storage enabled for ethash caches   dir=/tmp/gethdd/geth/ethash count=3
INFO [03-03|11:13:14.043] Disk storage enabled for ethash DAGs     dir=/home/ia/.ethash        count=2                                               
INFO [03-03|11:13:14.044] Initialising Ethereum protocol           versions="[65 64 63]" network=7 dbversion=7              
INFO [03-03|11:13:14.044] Loaded most recent local header          number=501957 hash=fc60cc…1aff3e td=392866616646 age=2mo2w4d                         
INFO [03-03|11:13:14.044] Loaded most recent local full block      number=0      hash=a68ebd…59a3f1 td=131072       age=5mo1d18h                          INFO [03-03|11:13:14.044] Loaded most recent local fast block      number=0      hash=a68ebd…59a3f1 td=131072       age=5mo1d18h
WARN [03-03|11:13:22.487] Truncating freezer table                 database=/tmp/gethdd/geth/chaindata/ancient table=bodies   items=501958 limit=1   
WARN [03-03|11:13:22.487] Truncating freezer table                 database=/tmp/gethdd/geth/chaindata/ancient table=receipts items=501958 limit=1        WARN [03-03|11:13:22.488] Truncating freezer table                 database=/tmp/gethdd/geth/chaindata/ancient table=diffs    items=501958 limit=1
WARN [03-03|11:13:22.488] Truncating freezer table                 database=/tmp/gethdd/geth/chaindata/ancient table=headers  items=501958 limit=1        
WARN [03-03|11:13:22.500] Truncating freezer table                 database=/tmp/gethdd/geth/chaindata/ancient table=hashes   items=501958 limit=1        INFO [03-03|11:13:22.502] Rewind ancient data                      number=0                                                                               
WARN [03-03|11:13:22.502] Truncate ancient chain                   from=501957 to=0                                                                  
DEBUG[03-03|11:13:22.503] Reinjecting stale transactions           count=0                                                                                
INFO [03-03|11:13:22.503] Loaded local transaction journal         transactions=0 dropped=0                                                               INFO [03-03|11:13:22.503] Regenerated local transaction journal    transactions=0 accounts=0                                                              
INFO [03-03|11:13:22.759] Allocated fast sync bloom                size=2.00GiB                                                                      
DEBUG[03-03|11:13:22.762] Recalculated downloader QoS values       rtt=20s confidence=1.000 ttl=1m0s                                                      DEBUG[03-03|11:13:22.786] UDP listener up                          addr=[::]:30303                                                                        
DEBUG[03-03|11:13:22.786] TCP listener up                          addr=[::]:30303                                                                   
INFO [03-03|11:13:22.800] New local node record                    seq=18 id=e99b48d753d5e698 ip=24.207.216.46 udp=30303 tcp=30303                        DEBUG[03-03|11:13:22.800] IP exceeds bucket limit                  ip=51.158.190.99                      
INFO [03-03|11:13:22.800] Started P2P networking                   self=enode://1ed8635e04ab7b9e242ee2c7298d409d09630a1049aff10d110f28042183e2ca0a1cbcbdb2
d6a5c3be79deb2b689d61009555f828e43f293082ba39b2a3986f6@24.207.216.46:30303                                                                             
DEBUG[03-03|11:13:22.800] InProc registered                        namespace=admin                                                                        
DEBUG[03-03|11:13:22.800] InProc registered                        namespace=admin       
```

**Comment by meowsbits** [comment]
The database(s) in this example (with [the database I'm using to reproduce the issue](http://isaacardis.com/etc/chaindata.bak/)) are eventually getting truncated to effective 0. Not sure if feature or bug.

**Comment by meowsbits** [comment]
It was a bug. 

Fixed with https://github.com/etclabscore/core-geth/pull/51/commits/c432502af9f82a7cd7dac1bb2624485c109d0257:

```
> rm -rf /tmp/gethdd/geth/chaindata && cp -a /tmp/gethdd/geth/chaindata.bak /tmp/gethdd/geth/chaindata && make geth && ./build/bin/geth --datadir /tmp/get
hdd --mordor --verbosity 4                                                                                                                                env GO111MODULE=on go run build/ci.go install ./cmd/geth                                                                                                 
>>> /home/ia/go1.13.5.linux-amd64/go/bin/go install -ldflags -X main.gitCommit=c432502af9f82a7cd7dac1bb2624485c109d0257 -X main.gitDate=20200303 -v ./cmd/geth                     
Done building.                                                                                        
Run "./build/bin/geth" to launch geth.                                                                
INFO [03-03|12:39:55.400] Bumping default cache on mainnet         provided=1024 updated=4096                 
DEBUG[03-03|12:39:55.400] Sanitizing Go's GC trigger               percent=25                                                  
INFO [03-03|12:39:55.402] Maximum peer count                       ETH=50 LES=0 total=50                                        
INFO [03-03|12:39:55.402] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory" 
DEBUG[03-03|12:39:55.402] FS scan times                            list=23.14µs set=714ns diff=897ns
INFO [03-03|12:39:55.404] Starting peer-to-peer node               instance=CoreGeth/v1.11.1-unstable-c432502a-20200303/linux-amd64/go1.13.5
INFO [03-03|12:39:55.404] Allocated trie memory caches             clean=1024.00MiB dirty=1024.00MiB
INFO [03-03|12:39:55.404] Allocated cache and file handles         database=/tmp/gethdd/geth/chaindata cache=2.00GiB handles=524288
DEBUG[03-03|12:39:55.612] Chain freezer table opened               database=/tmp/gethdd/geth/chaindata/ancient table=headers items=501958 size=134.81MiB
DEBUG[03-03|12:39:55.629] Chain freezer table opened               database=/tmp/gethdd/geth/chaindata/ancient table=hashes  items=501958 size=15.32MiB
DEBUG[03-03|12:39:55.640] Chain freezer table opened               database=/tmp/gethdd/geth/chaindata/ancient table=bodies  items=501958 size=2.75MiB
DEBUG[03-03|12:39:55.648] Chain freezer table opened               database=/tmp/gethdd/geth/chaindata/ancient table=receipts items=501958 size=1.48MiB
DEBUG[03-03|12:39:55.658] Chain freezer table opened               database=/tmp/gethdd/geth/chaindata/ancient table=diffs    items=501958 size=2.86MiB
INFO [03-03|12:39:55.658] Opened ancient database                  database=/tmp/gethdd/geth/chaindata/ancient                                            WARN [03-03|12:39:55.664] Freezer/KV validation error, attempting freezer repair error="gap (chaindb=#839245 frozen=#501958) in the chain between ancients
 and leveldb"                                                                     
WARN [03-03|12:39:55.664] Freezer repair OK                                       
WARN [03-03|12:39:55.664] Head header                              number=839245 hash=f2d626…e810c9
WARN [03-03|12:39:55.664] Head fast                                number=839245 hash=f2d626…e810c9
WARN [03-03|12:39:55.664] Head full                                number=0      hash=a68ebd…59a3f1
WARN [03-03|12:39:55.664] Persistent Freezer/KV gap: Truncating KV database to freezer height ancients=501958 kv.head_header_number=839245 kv.head_header_
hash=f2d626…e810c9                                                              
WARN [03-03|12:39:56.098] Removing KV block data                   n=830000 hash=0xa6c400d4ac4ef22addf56575cf79c2082577a00988e70c6e19c76b643baf0ba6
WARN [03-03|12:39:56.550] Removing KV block data                   n=820000 hash=0xdf812f64b9068c12fe1b83dd78e985433808e24b9ecab891166edf123ef8063d
WARN [03-03|12:39:56.986] Removing KV block data                   n=810000 hash=0xd925bb9d7ef68bc774c33026d3d033bd49038cd4ebc0eeacdfa278877e983856
WARN [03-03|12:39:57.440] Removing KV block data                   n=800000 hash=0xdd61074fdddd0ba9fa5c59ebf5e25ea807329a07587fda567ea5be8581b0a60a
WARN [03-03|12:39:57.903] Removing KV block data                   n=790000 hash=0x252ffed8c274a0980d49422b70b9fe7ec985ab12ef74212b09a20a5d2d89b10f
WARN [03-03|12:39:58.356] Removing KV block data                   n=780000 hash=0x7dbacff27c2e3b2dd305b3264e6e38c3ea4375025d852da99e0f9565aa367f82
WARN [03-03|12:39:58.836] Removing KV block data                   n=770000 hash=0xaba5c90a9c902967829ca05551d21a974c336740fe9d5191bb73f29e66018cee
WARN [03-03|12:39:59.315] Removing KV block data                   n=760000 hash=0x37125451ed0b0d2e06824ccfd52b04ed3b4b56a2403b1cc243dc180d5bf6e79a
WARN [03-03|12:39:59.795] Removing KV block data                   n=750000 hash=0x4b7a337d49ba8c3fa40fae4b1b51ffa101bc578615c0633b1e36c21c7e6fd41e
WARN [03-03|12:40:00.274] Removing KV block data                   n=740000 hash=0x0910ae3c6ea161148bb835bb4ce3fbea53f63759bced7d93b3cc984e8b9fd621
WARN [03-03|12:40:00.656] Removing KV block data                   n=730000 hash=0xe73117529fded7405786a22ef4bc7ee6359a2e54cd361353ce84f81ae0f3645f
WARN [03-03|12:40:01.038] Removing KV block data                   n=720000 hash=0x05c11768b81f00340cea324c7e7f662a48f8946a83c62ab24497e57212b6eb45
WARN [03-03|12:40:01.392] Removing KV block data                   n=710000 hash=0x5e2c9f44565bf028ebd9fde8346d1bfa30ec1dd2cf13313ccdbb768787519678
WARN [03-03|12:40:01.769] Removing KV block data                   n=700000 hash=0xcf5987b48846bf92802e35d9c435d915ceeeb5a68b7bab5b884881b05a0921dd
WARN [03-03|12:40:02.136] Removing KV block data                   n=690000 hash=0xb6f8baacee4f564029aff5638a09b54797f15a24b3b1540dd3c64440e4bf093b
WARN [03-03|12:40:02.486] Removing KV block data                   n=680000 hash=0x81e6f3be9ff819dbcdd80d7ac5fc73c7a7b81dbd5573f3d71699f74d52131523
WARN [03-03|12:40:02.843] Removing KV block data                   n=670000 hash=0xff3c2b100edc55ea00d1e29741772b9a36c244b09219742387715ec376f63924
WARN [03-03|12:40:03.191] Removing KV block data                   n=660000 hash=0x09bca32a68c46f0d3c90d83d648badbae5a04bb6931afd5fd89da0ccbc64043e
WARN [03-03|12:40:03.512] Removing KV block data                   n=650000 hash=0xc4cdd16ff8ee35697137a622966ef8365ea6d6232c4b3990afc93975f2236ef9
WARN [03-03|12:40:03.838] Removing KV block data                   n=640000 hash=0x391cd485c6430504698dc990e57bce9f4cd250bf1ba24ba21181ad919e319cf2
WARN [03-03|12:40:04.166] Removing KV block data                   n=630000 hash=0x2eed934410f220aaba8ed36b0c9d17ffed52ede3efd39f9a11a6f4e75dd45158
WARN [03-03|12:40:04.507] Removing KV block data                   n=620000 hash=0x8b3dfc97633c770d784874e8de5219c8a8e7c4232e1ace81f9f40e840326b5d9
WARN [03-03|12:40:04.844] Removing KV block data                   n=610000 hash=0xcbdd21b6e79dd4b4e844661aa0ffa6ea864aca03ac50e2c066d3c694419ef55d
WARN [03-03|12:40:05.190] Removing KV block data                   n=600000 hash=0x4c41542fb189ec3bf303a772f5e634073401069b9d1b21eaf73e495e029b5f9a
WARN [03-03|12:40:05.511] Removing KV block data                   n=590000 hash=0x774ce3ffc9ad6fb4ef8675bdf45ab20160c8ae1a1cf01659fc7443ee8805a56b
WARN [03-03|12:40:05.852] Removing KV block data                   n=580000 hash=0x393a8b1f394d09f06022053e912f5974986befd3ff4a8a2b368e144a0120bac0
WARN [03-03|12:40:06.187] Removing KV block data                   n=570000 hash=0xf3594daf24eb4c6da64b3e19a4a1e12a3e6327cde09ea8fe9583943e778e0d8a
WARN [03-03|12:40:06.500] Removing KV block data                   n=560000 hash=0x80e8a7e4895f577e7d0bbb7d329c831c215a23330edc3255237df6f2dab6e4a3
WARN [03-03|12:40:06.812] Removing KV block data                   n=550000 hash=0x42fdff6f256c1b412bb4d76e06569dd53d7eba18141f38169baf36b602932bb1
WARN [03-03|12:40:06.993] Finished KV truncation                                   
WARN [03-03|12:40:06.993] Writing KV head header                   hash=0xfc60ccf6933322f1bf8f81b570a3513ac21ed23e611e3388e7b0472d9f1aff3e
INFO [03-03|12:40:06.993] Persisted trie from memory database      nodes=0 size=0.00B     time=2.185µs gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesi
ze=0.00B                                                                        
DEBUG[03-03|12:40:06.993] Current full block not old enough        number=0      hash=a68ebd…59a3f1 delay=90000
INFO [03-03|12:40:06.993] Found stored genesis block               config="NetworkID: 7, ChainID: 63 Engine: ethash EIP1014: 301243 EIP1052: 301243 EIP140
: 0 EIP145: 301243 EIP150: 0 EIP155: 0 EIP160: 0 EIP161abc: 0 EIP161d: 0 EIP170: 0 EIP198: 0 EIP211: 0 EIP212: 0 EIP213: 0 EIP214: 0 EIP658: 0 EIP7: 0 Eth
ashECIP1017: 0 EthashECIP1041: 0 EthashEIP100B: 0 EthashEIP2: 0 EthashHomestead: 0 "
INFO [03-03|12:40:06.994] Initialised chain configuration          config="NetworkID: 7, ChainID: 63 Engine: ethash EIP1014: 301243 EIP1052: 301243 EIP110
8: 999983 EIP1344: 999983 EIP140: 0 EIP145: 301243 EIP150: 0 EIP152: 999983 EIP155: 0 EIP160: 0 EIP161abc: 0 EIP161d: 0 EIP170: 0 EIP1884: 999983 EIP198:
0 EIP2028: 999983 EIP211: 0 EIP212: 0 EIP213: 0 EIP214: 0 EIP2200: 999983 EIP658: 0 EIP7: 0 EthashECIP1017: 0 EthashECIP1041: 0 EthashEIP100B: 0 EthashEIP
2: 0 EthashHomestead: 0 "                                                                                                              
INFO [03-03|12:40:06.994] Disk storage enabled for ethash caches   dir=/tmp/gethdd/geth/ethash count=3                                 
INFO [03-03|12:40:06.994] Disk storage enabled for ethash DAGs     dir=/home/ia/.ethash        count=2
INFO [03-03|12:40:06.994] Initialising Ethereum protocol           versions="[65 64 63]" network=7 dbversion=7
INFO [03-03|12:40:06.995] Loaded most recent local header          number=501957 hash=fc60cc…1aff3e td=392866616646 age=2mo2w4d
INFO [03-03|12:40:06.995] Loaded most recent local full block      number=0      hash=a68ebd…59a3f1 td=131072       age=5mo1d20h
INFO [03-03|12:40:06.995] Loaded most recent local fast block      number=501957 hash=fc60cc…1aff3e td=392866616646 age=2mo2w4d
DEBUG[03-03|12:40:06.995] Reinjecting stale transactions           count=0
INFO [03-03|12:40:06.995] Loaded local transaction journal         transactions=0 dropped=0
INFO [03-03|12:40:06.995] Regenerated local transaction journal    transactions=0 accounts=0
INFO [03-03|12:40:07.229] Allocated fast sync bloom                size=2.00GiB
DEBUG[03-03|12:40:07.233] Recalculated downloader QoS values       rtt=20s confidence=1.000 ttl=1m0s             
DEBUG[03-03|12:40:07.257] UDP listener up                          addr=[::]:30303                                              
DEBUG[03-03|12:40:07.257] TCP listener up                          addr=[::]:30303                                                     
INFO [03-03|12:40:07.267] New local node record                    seq=22 id=e99b48d753d5e698 ip=24.207.216.46 udp=30303 tcp=30303
DEBUG[03-03|12:40:07.267] IP exceeds bucket limit                  ip=51.158.190.99                                                    
INFO [03-03|12:40:07.268] Started P2P networking                   self=enode://1ed8635e04ab7b9e242ee2c7298d409d09630a1049aff10d110f28042183e2ca0a1cbcbdb2
d6a5c3be79deb2b689d61009555f828e43f293082ba39b2a3986f6@24.207.216.46:30303
DEBUG[03-03|12:40:07.268] InProc registered                        namespace=admin
DEBUG[03-03|12:40:07.268] InProc registered                        namespace=admin
DEBUG[03-03|12:40:07.268] InProc registered                        namespace=debug
```

File: core/rawdb/database.go [hunk 1]
```diff
@@ -129,48 +130,150 @@ func NewDatabaseWithFreezer(db ethdb.KeyValueStore, freezer string, namespace st
...
-			// If the freezer is empty, ensure nothing was moved yet from the key-value
-			// store, otherwise we'll end up missing data. We check block #1 to decide
-			// if we froze anything previously or not, but do take care of databases with
-			// only the genesis block.
-			if ReadHeadHeaderHash(db) != common.BytesToHash(kvgenesis) {
-				// Key-value store contains more data than the genesis block, make sure we
-				// didn't freeze anything yet.
-				if kvblob, _ := db.Get(headerHashKey(1)); len(kvblob) == 0 {
-					return nil, errors.New("ancient chain segments already extracted, please set --datadir.ancient to the correct path")
+		}(), ReadHeadHeaderHash(ddb)
```
Comment (tzdybal) [inline core/rawdb/database.go]: For me this is extra over-complicated part. After reading the whole function, I need to go back and check were it's assigned. And There is a waste of `ReadHeadHeaderHash` call with such style.
We could first call `head_header_hash = ReadHeadHeaderHash(ddb)` and then use this value in `ReadHeaderNumber` few lines above, and save some IO operations.

File: core/rawdb/database.go [hunk 2]
```diff
@@ -129,48 +130,150 @@ func NewDatabaseWithFreezer(db ethdb.KeyValueStore, freezer string, namespace st
...
+		}
+		return nil
+	}
+
+	truncateKVtoFreezer := func(freezerdb *freezer, ddb ethdb.KeyValueStore) {
+		hhh := ReadHeadHeaderHash(ddb)
+		n := *ReadHeaderNumber(ddb, hhh)
+		frozen, _ := freezerdb.Ancients()
+
+		head_header, head_header_hash := func() uint64 {
```
Comment (tzdybal) [inline core/rawdb/database.go]: I really don't understand why you're using anonymous `func` in here. "Normal" code would be absolutely sufficient, and IMHO more readable.

Comment (tzdybal) [inline core/rawdb/database.go]: And variable names are breaking the Go naming convention (see: [Effective Go](https://golang.org/doc/effective_go.html#mixed-caps)).

File: core/rawdb/database.go [hunk 3]
```diff
@@ -129,48 +130,150 @@ func NewDatabaseWithFreezer(db ethdb.KeyValueStore, freezer string, namespace st
...
-				if kvblob, _ := db.Get(headerHashKey(1)); len(kvblob) == 0 {
-					return nil, errors.New("ancient chain segments already extracted, please set --datadir.ancient to the correct path")
+		}(), ReadHeadHeaderHash(ddb)
+		head_fast, head_fast_hash := func() uint64 {
+			if n := ReadHeaderNumber(ddb, ReadHeadFastBlockHash(ddb)); n == nil {
+				return 0
+			} else {
+				return *n
+			}
+		}(), ReadHeadFastBlockHash(ddb)
```
Comment (tzdybal) [inline core/rawdb/database.go]: Same here.

File: core/rawdb/database.go [hunk 4]
```diff
@@ -129,48 +130,150 @@ func NewDatabaseWithFreezer(db ethdb.KeyValueStore, freezer string, namespace st
...
+		head_full, head_full_hash := func() uint64 {
+			if n := ReadHeaderNumber(ddb, ReadHeadBlockHash(ddb)); n == nil {
+				return 0
+			} else {
+				return *n
+			}
+		}(), ReadHeadBlockHash(ddb)
```
Comment (tzdybal) [inline core/rawdb/database.go]: And here....

**Comment by meowsbits** [comment]
@tzdybal Any improvements, nits, or edits please just go ahead and push 'em right to the branch!

File: core/rawdb/database.go [hunk 5]
```diff
@@ -129,48 +130,150 @@ func NewDatabaseWithFreezer(db ethdb.KeyValueStore, freezer string, namespace st
...
-				return nil, fmt.Errorf("genesis mismatch: %#x (leveldb) != %#x (ancients)", kvgenesis, frgenesis)
-			}
-			// Key-value store and freezer belong to the same network. Ensure that they
-			// are contiguous, otherwise we might end up with a non-functional freezer.
-			if kvhash, _ := db.Get(headerHashKey(frozen)); len(kvhash) == 0 {
-				// Subsequent header after the freezer limit is missing from the database.
-				// Reject startup is the database has a more recent head.
-				if *ReadHeaderNumber(db, ReadHeadHeaderHash(db)) > frozen-1 {
-					return nil, fmt.Errorf("gap (#%d) in the chain between ancients and leveldb", frozen)
+	validateFreezerVsKV := func(freezerdb *freezer, ddb ethdb.KeyValueStore) error {
```
Comment (tzdybal) [inline core/rawdb/database.go]: IMHO this should be refactored as a separate function, not lambda.

File: core/rawdb/database.go [hunk 6]
```diff
@@ -129,48 +130,150 @@ func NewDatabaseWithFreezer(db ethdb.KeyValueStore, freezer string, namespace st
...
+		// If the genesis hash is empty, we have a new key-value store, so nothing to
+		// validate in this method. If, however, the genesis hash is not nil, compare
+		// it to the freezer content.
+		if kvgenesis, _ := db.Get(headerHashKey(0)); len(kvgenesis) > 0 {
+			if frozen, _ := frdb.Ancients(); frozen > 0 {
```
Comment (tzdybal) [inline core/rawdb/database.go]: Why use `frdb` if we're passing `freezerdb` as parameter?
Why pass `freezerdb` as parameter if we're using `frdb` anyways?
;)

**Review by tzdybal — CHANGES_REQUESTED** [review]
I agree that `NewDatabaseWithFreezer` is the right place to fix this problem. And I agree that rewinding is the only hope to fix the "gap" problem.

I don't like the complexity of the code - some parts are hard to understand and IMHO there is extra complexity, that is easy to avoid (like lambdas executed only once, assignment to two vars in single line, where first value is result of a inline-defined anonymous func). 

I made PR to this branch with my proposed refactoring.

**Comment by meowsbits** [comment]
Yeap I like your refactorings. LGTM.

**Comment by meowsbits** [comment]
Just a couple last comments added and newlines adjusted. And made an error (the one from `freezer#repair`) not go silent into the dark night.

**Review by tzdybal — APPROVED** [review]

**Pull request merged**